### ISSUE TITLE
PinballY gets confused with table names with apostrophies

### ISSUE THREAD

**Issue by rtl19** [mjrgh/PinballY#108]
I thought I was going crazy.  I went through each table and made sure that media was being displayed.  In cases where names like "X's and O's" or "America's Most Haunted", they were getting displayed in the info window as "America&apos;s Most Haunted" so I would go into the edit details screen and fix it.  On exiting the details screen it displayed fine.  But when I cycled the wheel back to that game or left PinballY and came back in, it was messed up.  Not always but sometimes.

Just 2 seconds ago I corrected America&apos;s Most Wanted again back to the apostrophy and it didn't display correctly after cycling the wheel and returning.

**Comment by rtl19** [comment]
Just a note of clarification that I think the github editor masked out  when i typed it ... the apostrophies are getting displayed as the ampersand apos semi-colon notation.

**Comment by mjrgh** [comment]
Could you give me a step-by-step here?  I don't see a problem when I try it - I want to make sure I'm trying the same thing.

**Comment by rtl19** [comment]
Sure.  I'll try to recreate it.

First, I'll look in the database XML file (Visual Pinball.xml).  I see this:
`"Americas Most Wanted".  Here is the paraphrased xml entry for that table:

>   Game name="Americas Most Haunted (spooky 2014)"
 >                 Description=America&amp;apos;s Most Haunted (Original 2014)

When displayed in Pinball Y, it shows the amp;apos text.
Now I go into 'Game Setup', then 'Edit Details'
I got into the Title drop down arrow and pick IMDB entry 6161.
Click Ok or Save.  It displays correctly.

Now I cycle thru some other games using the flipper buttons.
I pick another game with a name without apostrophies and edit those details as well.  Just for fun :)
I save that info and cycle through a few more games.  Then I go back to America's Most Wanted.  
So far it looks fine.

**Comment by rtl19** [comment]
I must have closed this accidentally.  Sorry about that and the weird font sizing.

**Comment by mjrgh** [comment]
Let's suppose you're starting from scratch **without** any XML files.  Can you make this happen by entering the name **entirely from within PinballY**?

**Comment by MikePinball** [comment]
I see this problem too. In my case I still use a shared database with PinballX. I get it all setup nicely in PinballX with an apostrophe. Then when I open it in PinballY it becomes &apos. Even if I edit it back correctly in PinballY, it later shows up as &apos in the menu. I'm guessing that whenver you edit it a table, it doesn't properly change back the @apos. to a "'".

**Comment by mjrgh** [comment]
There's got to be something you guys are doing that you aren't telling me - I'm not seeing it.  Maybe if you could lead me through it even more step by step?  I tried the exact steps you listed and everything's looking right to me.

**Comment by mjrgh** [comment]
And rtl19, maybe you could try clarifying exactly what you're seeing.  When I read your initial note here, here's what I see (paraphrasing):

> In cases where names like "America's Most Haunted", they were getting displayed in the info window as "America's Most Haunted"

Which looks you're saying "Everything looks right here!".  Which I assume isn't what you mean - you said something about github changing around your text.  Maybe you could try writing it out in some other notation if you can't get the characters to look right, or post a picture or something.  I'm a little confused about what you're seeing.

**Comment by rtl19** [comment]
> Let's suppose you're starting from scratch **without** any XML files. Can you make this happen by entering the name **entirely from within PinballY**?

Yes.  I'll do that today.  I'll also work on the more detailed step-by-step.   I'll try to get screen captures as well.  By the way, if remoting into my pinball box is worthwhile for you, I can set that up as well.

**Comment by rtl19** [comment]
Ok.  I'm attching a zip file of screen captures that demonstrates the steps I took originally.  As in before wiping the xml clean.  One interesting thing to note is I looked in the Gamestats.csv file and there are 2 entries for America's Most Haunted.  Here are their entries:

> FirePower (Williams 1980).Visual Pinball,,,,,,,,Yes
"America's Most Haunted (Spooky Pinball LLC, 2014) (Spooky Pinball LLC 2014).Visual Pinball - X",,,,,,,,,20191222050000,Auto,,
Airborne (Capcom 1996).Visual Pinball,,,,,,,,,20191222050000,Auto,,
America's Most Haunted (Spooky Pinball LLC 2014).Visual Pinball - X,,,,,,,,,20191222050000,Auto,,

I'll post the fresh start version of the screen captures soon.  I'm 
[ApostrophyScreenCaptures.zip](https://github.com/mjrgh/PinballY/files/4081204/ApostrophyScreenCaptures.zip)
assuming I can just delete the entries for America's Most Haunted and not have to wipe the entire xml.   If you prefer I wipe the entire xml, let me know.

**Comment by rtl19** [comment]
I'm attaching the post XML experiment.  As you probably suspected, it worked fine.  I'll have to fire up Pinball X to see if the entry that Pin Y put in there will play nice with X.  
[ApostrophyCaptures_afterXML.zip](https://github.com/mjrgh/PinballY/files/4081329/ApostrophyCaptures_afterXML.zip)

**Comment by mjrgh** [comment]
Thanks for the screen shots - that's very helpful.  I think I see what's been going on now.

Looking at your original XML (before_1_xmldatabase.jpg), here's what I see:

`America&amp;apos;s Most Haunted (Original 2014)`

If you carefully parse this XML code, here's what it's saying in XML-speak:  "display the letters 'America', then display a literal ampersand, then the letters 'apos', then a literal semicolon, then a space, then the letters 'Most Haunted (Original 2014)'".  So when PinballY displays this, it's correctly interpreting the XML when it displays "America" followed by an ampersand followed by the letters "apos" and so on.  That's presumably not what you wanted, and is the whole reason for this issue!, but it's what your XML actually instructs PinballY to do.  If you ran this XML through any other XML interpreter that's reading the code correctly, it would do the same thing.  Certainly what you really wanted to be displayed was the string "America's Most Haunted (Original 2014)".  To get that, the XML code that **should** be in the file is

`America&apos;s Most Haunted (Original 2014)`

So my first question is: what does the XML file look like after you exited PinballY in step 8?  PinballY **should** have updated the file so that the XML looks like the above.  If it didn't, let's figure out why.

My second question is: assuming that the XML looked right after step 8, did you do anything outside of PinballY between steps 8 and 9 that would have changed the XML back to the faulty original format?

Third question is where did the original faulty XML came from in the first place?  Is that from PinballX, maybe?

**Comment by rtl19** [comment]
To answer your first question, take a look at the attachment titled: postxml_10_xmldata.jpg.  Basically it looks like what you hope to see.  I'm not following your second question...I didn't change anything after step 8.  The xml is not back to the faulty format.  Why do you think it's back to the faulty format? 

Third question...prior to experimenting, and enjoying, Pinball Y, I've only ever used Pinball X.  The XML was created using the Game Manager tool that comes with Pinball X.  It's quite possible the table name that I download some years ago had the funky ampersand stuff in the title so that when I copied that into Game Manager it did something funky.  I'm not totally sure.

I have a few other games with apostrophes and weird displays like "X's and O's" and "Bram Stoker's Dracula".  I'm going to fix the filenames (if needed) and remove their entries from the xml and let Pinball Y discover them.  Hopefully that will take care of it.

**Comment by mjrgh** [comment]
> I'm not following your second question...I didn't change anything after step 8. The xml is not back to the faulty format. Why do you think it's back to the faulty format?

That came from looking at after_9_returntopix.jpg - it looked like the "apos" string was back, so I thought the original XML with the extra ampersand-"amp;apos;" string was back.

> The XML was created using the Game Manager tool that comes with Pinball X. It's quite possible the table name that I download some years ago had the funky ampersand stuff in the title so that when I copied that into Game Manager it did something funky. I'm not totally sure.

Okay, that makes sense.  Let's see...

Yep!  That's definitely it.  I just tried running the PBX Game Manager with an XML file containing a title with an apostrophe, and it writes it back with the screwy `&amp;apos;` data.  So there's the source.  I guess you just have to be careful not to use the PBX Game Manager with shared files. (Presumably PinballX itself won't rewrite the XML files, though, so hopefully it's safe to run that on a shared configuration.)

> I have a few other games with apostrophes and weird displays like "X's and O's" and "Bram Stoker's Dracula". I'm going to fix the filenames (if needed) and remove their entries from the xml and let Pinball Y discover them. Hopefully that will take care of it.

Sounds good - let me know how that goes.

**Comment by mjrgh** [comment]
Okay, I've changed it for the next update so that PinballY won't do the `&apos;` expansions - that should work around the PBX limitation and make file sharing more compatible.

**Comment by rtl19** [comment]
Cool.  One question I have from the previous behavior... when I edited the game details on one of the games, why did it work after that until I exited PinY?

**Comment by mjrgh** [comment]
> One question I have from the previous behavior... when I edited the game details on one of the games, why did it work after that until I exited PinY?

Unclear - that's basically what I was asking in the "second question" earlier, isn't it?  Can you reproduce the sequence of events that gets you there?  If so, what does the XML look like after the step where you exit PinballY?  I'd think the XML must be screwed up again, so the question is how did it get screwed up again?

**Comment by rtl19** [comment]
My first zip of attachments is my reproduction of what got me to the problem point.  It seems like PinY wasn't saving whatever it needed to save in the right way to get around the original issue.  I guess we'll never know.

I think my approach of removing the offending lines from the xml and having PinY recreate those lines is working.  I haven't tried going into Pin X's game manager to see if saving there causes the corruption.

**Comment by mjrgh** [comment]
> My first zip of attachments is my reproduction of what got me to the problem point. It seems like PinY wasn't saving whatever it needed to save in the right way to get around the original issue. I guess we'll never know.

I agree, it'll remain a mystery unless you can reproduce that sequence again.  If you can, let me know and we can look into it further.

>  I haven't tried going into Pin X's game manager to see if saving there causes the corruption.

It will - that's reproducible on my end with the PBX Game Manager.  If a file contains the `&apos;` sequence, and you load it into PBX Game Manager, PBX Game Manager will rewrite it as `&amp;apos;` on saving the file.  It's a good bet that's where all of this was coming from originally.

**Comment by mjrgh** [comment]
The change to remove the `&apos;` expansion is in the Beta 19, and since we seem to know the root cause now, I'm going to call this done.
